### ISSUE TITLE
enumerate regression when doing an "or []" inside it on 1.1.308

### ISSUE THREAD

**Issue by bellini666** [microsoft/pyright#5106]
```python
def foo1(x: list[str] | None):
    for i, v in enumerate(x or []):
        reveal_type(v)


def foo2(x: list[str] | None):
    if not x:
        x = []
    for i, v in enumerate(x):
        reveal_type(v)


def foo3(x: list[str] | None):
    if x is None:
        x = []
    for i, v in enumerate(x):
        reveal_type(v)


def foo4(x: list[str] | None):
    x = x or []
    for i, v in enumerate(x):
        reveal_type(v)
```

Running this produces:

```
❯ pyright test.py 
/tmp/test.py
  /tmp/test.py:3:21 - information: Type of "v" is "_T@enumerate"
  /tmp/test.py:10:21 - information: Type of "v" is "str"
  /tmp/test.py:17:21 - information: Type of "v" is "str"
  /tmp/test.py:23:21 - information: Type of "v" is "str"
0 errors, 0 warnings, 4 informations
```

All those functions are logically the same, but `foo1` started behaving differently after 1.1.308 (before it would behave the same as the others, revealing the correct type)

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.309, which I just published. It will also be included in a future release of pylance.
